File: programs/program_05.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int *array;
    int top;
    int capacity;
} stack;

stack *createStack(int capacity) {
    stack *s = (stack *)malloc(sizeof(stack));
    s->array = (int *)(malloc(sizeof(int) * capacity));
    s->capacity = capacity;
    s->top = -1;
}

void pop(stack *s) {
    if (s->top != -1) {
        s->top--;
        printf("Top element popped!\n");
    } else
        printf("ERROR : Stack underflow");
}

int isFull(stack *s) {
    if (s->top == s->capacity - 1)
        return 1;
    return 0;
}

int isEmpty(stack *s) {
    if (s->top == -1)
        return 1;
    return 0;
}

void push(stack *s, int x) {
    if (s->top != s->capacity - 1) {
        s->array[++s->top] = x;
        printf("Pushed %d!\n", x);
    } else
        printf("ERROR : Stack overflow");
}

int top(stack *s) {
    if (s->top != -1)
        return s->array[s->top];
}

int main() {
    stack *s = createStack(10);
    push(s, 10);
    push(s, 72);
    push(s, 45);
    printf("Top element : %d\n", top(s));
    pop(s);
    pop(s);
    printf("Top element : %d\n", top(s));
    return 0;
}

File: programs/program_05.rs
struct Stack<T> {
    stack: Vec<T>,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack { stack: Vec::new() }
    }
    fn push(&mut self, item: T) {
        self.stack.push(item);
    }
    fn pop(&mut self) -> Option<T> {
        self.stack.pop()
    }
    fn top(&self) -> &T {
        self.stack.last().expect("Stack is empty")
    }
    fn length(&self) -> usize {
        self.stack.len()
    }
    fn is_empty(&self) -> bool {
        self.stack.is_empty()
    }
}

fn main() {
    let mut s: Stack<i32> = Stack::new();
    println!("EMPTY : {}", s.is_empty());
    s.push(6);
    s.push(7);
    println!("{} {} {}", s.top(), s.length(), s.is_empty());
    s.push(8);
    println!("{} {} {}", s.top(), s.length(), s.is_empty());
    s.pop();
    println!("{} {} {}", s.top(), s.length(), s.is_empty());
}
